### PR TITLE
2nd phase of deprecation of "x ident" in constr notation (now an error)

### PR DESCRIPTION
**Pull request by herbelin**
**Kind:** deprecation

This is the 8.14 step of the transition of `x ident` to be changed into `x name` into (non custom) notations. See protocol in this [comment](https://github.com/coq/coq/pull/11841#issuecomment-728756462) of #11841.

### CODE REVIEW

**Comment by Zimmi48** [comment]
This should get a changelog entry. But this will also make it harder to backport for @silene. So if @silene accepts that this can go in 8.14, the changelog entry should in fact be directly inserted into the 8.14 release notes and then the PR should be merged quickly.

**Comment by herbelin** [comment]
> But this will also make it harder to backport for @silene.

I'm waiting for knowing the result of CI but this PR is indeed a question to @silene (with apologies for not having done it earlier).

**Comment by Zimmi48** [comment]
Making the warning fatal breaks a lot of projects in CI. So, it looks like this is a bit late for 8.14 and should be retargeted to 8.15.

**Comment by herbelin** [comment]
Yes, this is a lot of changes. That would require an automatic translation, but maybe should I renounce to the change and let `ident` and `name` be synonymous, and maybe add instead something like `strict ident`/`strict name`.

**Comment by Zimmi48** [comment]
> maybe add instead something like `strict ident`/`strict name`.

Providing `strict ident` could make sense as a transition step as it would:
- give a way to use the new meaning of `ident` without waiting for the end of the deprecation (which could take a long time given the number of projects affected);
- give a way to resolve the warning without disabling it for people who actually meant `ident` (and thus want the new behavior).

However, `strict name` doesn't really bring any benefit since `name` is already new and the two would mean the same.

**Comment by herbelin** [comment]
> > maybe add instead something like `strict ident`/`strict name`.
> 
> Providing `strict ident` could make sense as a transition step as it would:
> 
>     * give a way to use the new meaning of `ident` without waiting for the end of the deprecation (which could take a long time given the number of projects affected);

Who would make the changes? We could easily provide a tool but is it at the end really worth the burden?

> However, `strict name` doesn't really bring any benefit since `name` is already new and the two would mean the same.

Right.

**Comment by Zimmi48** [comment]
>Who would make the changes? We could easily provide a tool but is it at the end really worth the burden?

If your question is: is it worth the burden to deprecate `ident` if we already provide `strict ident`, then indeed, I'm not sure. The deprecation warning could just become a normal warning, advising that `ident` has the meaning of `name` and to use `strict ident` if the intent is not to include `_`.

**Comment by herbelin** [comment]
@Zimmi48 : Yes. Then I'll think about that.

**Review by Alizter — CHANGES_REQUESTED** [review]
LGTM modulo changelog and doc

**Comment by Alizter** [comment]
@herbelin ping

**Comment by Alizter** [comment]
@herbelin ping ping

**Comment by herbelin** [comment]
I'm still a bit hesitant. Addressing the `ident` warning and changing it to `name` will require some work but will eventually be nicer. So, I'm still in favor of keeping the deprecation and to give more time for the developments to be adapted. Then, later, in 2 years maybe, we can change the semantics of `ident`.

About `strict ident`, it is not clear whether if it is really necessary as we could live without it for years already and can wait a bit longer. I kept it so that even if the change finally does not happen, there is an alternative to really mean `ident`.

**Comment by herbelin** [comment]
In any case, I added a changelog and adapted the refman. Taking some more time to think about it.

File: doc/sphinx/user-extensions/syntax-extensions.rst
Comment (jfehrle) [inline doc/sphinx/user-extensions/syntax-extensions.rst:635]:
```suggestion
notation. It indicates :g:`x` should be parsed as a :n:`@name` (i.e., an
:n:`@ident` or the unnamed variable :g:`_`).
```

```suggestion
notation. It indicates to parse :g:`x` as a :n:`@name` (that is an
```

File: doc/sphinx/user-extensions/syntax-extensions.rst
Comment (jfehrle) [inline doc/sphinx/user-extensions/syntax-extensions.rst:989]:
```suggestion
          parse an :n:`@ident` excluding :g:`_` as ``strict ident`` does.
```

File: doc/sphinx/user-extensions/syntax-extensions.rst
Comment (jfehrle) [inline doc/sphinx/user-extensions/syntax-extensions.rst:1257]:
```suggestion
   ``global``
     Parses :n:`@qualid` (and thus, as a special case, :n:`@ident`).
     
   ``bigint``
      Parses arbitrary large integer numbers as in :n:`@bigint`.

   ``strict ident``
     Parses identifiers as in grammar entry
     :n:`@ident`. The entry ``name`` parses :n:`@name`, which allows
     both :n:`@ident`\s and :g:`_`.
```

The order of these items should be the same as in the grammar.  And move this above all the notes, around line 1217

File: doc/sphinx/user-extensions/syntax-extensions.rst
Comment (jfehrle) [inline doc/sphinx/user-extensions/syntax-extensions.rst:1261]:
```suggestion
   As of version 8.16, the entry ``ident`` is a deprecated alias for
```

If correct

File: doc/sphinx/user-extensions/syntax-extensions.rst
Comment (jfehrle) [inline doc/sphinx/user-extensions/syntax-extensions.rst:1265]:
```suggestion
   ``ident`` may be rebound to :n:`@ident`.  Note that this does not
   apply to custom entries, where ``ident`` already parses
```

File: doc/sphinx/user-extensions/syntax-extensions.rst
Comment (jfehrle) [inline doc/sphinx/user-extensions/syntax-extensions.rst:1268]:
```suggestion
   :n:`@ident`. Nor does it apply to binders parsed as
   :n:`@term` but interpreted as identifiers where the ``as ident``
   modifier already excludes :g:`_`.
```

"interpreted as identifiers" meaning it is a term that consists of a single \@ident?  Would you clarify?

**Review by jfehrle — COMMENTED** [review]
wording suggestions

File: doc/sphinx/user-extensions/syntax-extensions.rst
Comment (herbelin) [reply doc/sphinx/user-extensions/syntax-extensions.rst]: Yes, that's the point. I clarified.

File: doc/sphinx/user-extensions/syntax-extensions.rst
Comment (herbelin) [reply doc/sphinx/user-extensions/syntax-extensions.rst]: Both 8.13 and 8.15 are incorrect. The deprecation was actually in 8.14.

File: doc/sphinx/user-extensions/syntax-extensions.rst
Comment (herbelin) [reply doc/sphinx/user-extensions/syntax-extensions.rst]: OK, adding then a paragraph for `name` too I guess.

**Review by herbelin — COMMENTED** [review]
Thanks a lot for the comments. I added a new commit answering to the remarks. (Will need squashing at some time.)

**Comment by silene** [comment]
Did you mistakenly commit the `strbrk` changes into this pull request?

Also, I find the changelog entry quite confusing. Indeed, at no point does it explain what `strict ident` actually recognizes. The reader has to infer it from what the other constructs recognize. I suggest something along the lines of "New notation `strict ident`, which recognizes pure identifiers. This is meant to ease the transition away from `ident`. It offers an alternative to `name`, which recognizes both identifiers and `_`."

File: doc/changelog/03-notations/14637-master+notation-type-ident-deprecation-error-phase-8.14.rst
Comment (jfehrle) [inline doc/changelog/03-notations/14637-master+notation-type-ident-deprecation-error-phase-8.14.rst:4]:
```suggestion
  New notation ``strict ident`` in :n:`@explicit_subentry` that only accepts
  :n:`@ident`.  This is meant to ease the transition from
  the inaccurately named ``ident`` to ``name``.  Currently (except in custom
```

Following @silene's comment.

**Comment by herbelin** [comment]
Adding `strict ident` is not so nice. Let me try again to evaluate the cost of proposing a script for automatic upgrading.

**Pull request closed without merging**
